File: src/python/ops.py

# src/python/ops.py
import mlx.core as mx
# Import the C++ binding module built by CMake/Nanobind
# The name '_pal_cpp_binding' must match NB_MODULE in binding.cpp
# and the target name in CMakeLists.txt
from . import _pal_cpp_binding

def paged_attention(
    q: mx.array,
    kv_cache: mx.array, # Full KV cache buffer
    page_table: mx.array, # Logical block -> physical page mapping
    stream: mx.Stream | mx.Device | None = None # Optional stream/device
) -> mx.array:
    """
    Performs paged attention using the custom C++ primitive and Metal kernel.

    Args:
        q (mlx.core.array): Queries array. Shape depends on batching,
                            e.g., [batch_size, num_heads, seq_len, head_dim] or flattened.
        kv_cache (mlx.core.array): The entire KV cache buffer allocated by the PageAllocator.
                                 Shape e.g., [num_pages, tokens_per_page, 2, num_kv_heads, head_dim].
        page_table (mlx.core.array): Page table mapping logical blocks for each sequence
                                   to physical pages in kv_cache. Shape depends on kernel design.
        stream (mlx.core.Stream | mx.core.Device | None, optional): Stream or device to evaluate on.
                                                                    Defaults to None (MLX default).

    Returns:
        mlx.core.array: The attention output array. Shape likely matches 'q' or a derivative.
    """
    # --- Input Validation (Optional but Recommended) ---
    # Example checks (adapt based on actual kernel requirements)
    if q.dtype not in [mx.float16, mx.bfloat16, mx.float32]:
         raise TypeError(f"Query dtype {q.dtype} not supported.")
    if kv_cache.dtype != q.dtype:
         raise TypeError(f"KV cache dtype {kv_cache.dtype} must match query dtype {q.dtype}.")
    if page_table.dtype != mx.uint32:
         raise TypeError(f"Page table dtype must be uint32, got {page_table.dtype}.")
    # Add shape checks as needed

    # --- Call the C++ Bound Operation ---
    # Pass arguments directly. Nanobind handles type conversions.
    # The C++ operation `pal::cpp::paged_attention` creates the primitive
    # and returns the output array object, adding it to the MLX graph.
    output_array = _pal_cpp_binding.paged_attention(
        q,
        kv_cache,
        page_table,
        stream=stream # Pass the stream/device context
    )

    return output_array

File: src/cpp/ops/ops.cpp
#include "ops.hpp"
#include "paged_attention_primitive.hpp"
#include <mlx/ops.h>
#include <mlx/utils.h>
#include <stdexcept>
#include <vector>
#include <iostream>

namespace pal::cpp {

// Define the user-facing C++ operation function
mx::array paged_attention(
    const mx::array& q,
    const mx::array& kv_cache,
    const mx::array& page_table,
    const mx::Stream& s
) {
    std::cerr << "[Debug] pal::cpp::paged_attention C++ operation called" << std::endl;

    // --- Input Validation (Example) ---
    // Add more checks as needed based on kernel requirements
    if (q.ndim() < 2 || kv_cache.ndim() < 1 || page_table.ndim() < 1) {
         throw std::invalid_argument("[paged_attention] Invalid input dimensions.");
    }
     if (q.dtype() != mx::float16 || kv_cache.dtype() != mx::float16) {
         throw std::invalid_argument("[paged_attention] Stub requires float16 queries and KV cache.");
     }
     if (page_table.dtype() != mx::uint32) {
          throw std::invalid_argument("[paged_attention] Page table must be uint32.");
     }

    // --- Determine Output Shape ---
    // For the stub, assume output shape matches query shape.
    // A real implementation might calculate based on head concatenation, etc.
    auto out_shape = q.shape();
    auto out_dtype = q.dtype(); // Output dtype usually matches query/value dtype

    // --- Create Primitive Instance ---
    // Pass the stream/device context to the primitive
    auto primitive = std::make_shared<PagedAttentionPrimitive>(mx::to_stream(s));
    std::cerr << "[Debug] paged_attention: Created PagedAttentionPrimitive instance" << std::endl;


    // --- Construct Output Array ---
    // Create the output array object. This adds the operation to the graph.
    // The actual computation happens later when mx::eval() is called.
    return mx::array(
        out_shape,                        // Shape of the output
        out_dtype,                        // Dtype of the output
        primitive,                        // The primitive performing the computation
        {q, kv_cache, page_table}         // Inputs to the primitive
    );
}

} // namespace pal::cpp
